concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,149 @@
+            public_key=settings.SIGNING_SERVICE_PUBLIC_KEY,
+        )
+    # Received data frame is invalid or the signature does not match its content.
+    except MiddlemanProtocolError as exception:
+        raise SCICallbackFrameError() from exception
+    # If the received Golem message is malformed or not signed by the Signing Service.
+    except MessageError as exception:
+        raise SCICallbackPayloadError() from exception
+
+    # Request ID in the received data frame does not match the one in the data frame that was sent.
+    if deserialized_message.request_id != transaction.nonce:  # TODO: How to generate request_id ??
+        raise SCICallbackRequestIdError('MiddleMan response request_id does not match requested.')
+
+    # If the received Golem message is TransactionRejected or an ERROR frame.
+    if isinstance(deserialized_message, ErrorFrame) or isinstance(deserialized_message.payload, TransactionRejected):
+        raise SCICallbackPayloadError(
+            'Received frame is either ErrorFrame or contains Golem message TransactionRejected.'
+        )
+
+    # If the received Golem message is anything else than SignedTransaction, TransactionRejected or and ERROR frame.
+    if not isinstance(deserialized_message.payload, SignedTransaction):
+        raise SCICallbackPayloadError('Received frame payload is not Golem message SignedTransaction instance.')
+
+    signed_transaction = deserialized_message.payload
+
+    # The transaction data in SignedTransaction does not match the information we sent.
+    if not (
+        signed_transaction.nonce == transaction.nonce and
+        signed_transaction.gasprice == transaction.gasprice and
+        signed_transaction.startgas == transaction.startgas and
+        signed_transaction.to == transaction.to and
+        signed_transaction.value == transaction.value and
+        signed_transaction.data == transaction.data
+    ):
+        raise SCICallbackPayloadError('Received SignedTransaction does not match TransactionSigningRequest.')
+
+    # If the Ethereum signature in SignedTransaction is not a valid signature or
+    # or the transaction in SignedTransaction is not signed with the right Ethereum private key.
+    message_hash = sha3(encode(transaction, UnsignedTransaction))
+
+    if not ecdsa_raw_verify(
+        message_hash,
+        (
+            signed_transaction.v,
+            signed_transaction.r,
+            signed_transaction.s,
+        ),
+        settings.CONCENT_ETHEREUM_PUBLIC_KEY,
+    ):
+        raise SCICallbackSignatureError(
+            'Received SignedTransaction signature data is not signed by right Ethereum private key.'
+        )
+
+    # If the response is a valid SignedTransaction, copy the signature to the transaction object passed to the
+    # callback by SCI and return.
+
+    transaction.v = signed_transaction.v
+    transaction.r = signed_transaction.r
+    transaction.s = signed_transaction.s
+
+    return transaction
```

Comment:
hm... we are doing an interesting thing here:
1. we modify the object that was given as parameter
2. we return it
Maybe it would be better to make a copy of the `Transaction` object, modify it and return it?

Reply:
What would be benefit of it ?

Reply:
But we don't know if that is not expected either ;) Ok, done :+1: